### PR TITLE
sync translation of 'build custom form controls'

### LINKED ISSUE #14421: Some live sample in "How to build custom form controls" is broken
### MDN URL

https://developer.mozilla.org/zh-CN/docs/Learn/Forms/How_to_build_custom_form_controls

### 与此议题相关的小节或标题？

_No response_

### 哪些信息是错误、无用或不完善的？

See https://github.com/mdn/translated-content/pull/14293#issuecomment-1637064573, and mdn/content#27813. We should update this document to fix the build error and the broken live samples.

### 你期望看到什么？

The above errors are resolved.

### 你有什么支持的链接、参考或引用？

_No response_

### 你还想补充什么内容？

I would assigned myself to fix this :)

### MDN metadata

<details>
<summary>Page report details</summary>

* Folder: `zh-cn/learn/forms/how_to_build_custom_form_controls`
* MDN URL: https://developer.mozilla.org/zh-CN/docs/Learn/Forms/How_to_build_custom_form_controls
* GitHub URL: https://github.com/mdn/translated-content/blob/main/files/zh-cn/learn/forms/how_to_build_custom_form_controls/index.md
* Last commit: https://github.com/mdn/translated-content/commit/d70eee4b8cf68344a9d5544cf94d090494d49a66
* Document last modified: 2023-07-18T09:52:27.000Z

</details>

### PR DESCRIPTION
**Pull request by yin1999**
### Description

sync translation of 'build custom form controls'

### Related issues and pull requests

Fixes: #14421

Blocked by: mdn/content#28148, mdn/content#28167.

### CODE REVIEW

**Comment by yin1999** [comment]
Note for myself: update the related example code under `/zh-CN/docs/Learn/Forms/How_to_build_custom_form_controls/`

**Comment by yin1999** [comment]
As this is a "large" PR, @jasonren0403 do you have time to take a look in advance :)

File: files/zh-cn/learn/forms/how_to_build_custom_form_controls/index.md [hunk 1]
```diff
@@ -5,80 +5,87 @@ slug: Learn/Forms/How_to_build_custom_form_controls
...
 
-另外一个有趣的例子是：当小部件处于打开状态时，用户按下键盘上方向键和下方向键将会发生什么？这个问题有些棘手，如果你认为活动状态和打开状态是完全不同的，那么答案就是“什么都不会发生”，因为我们没有定义任何在打开状态下键盘的交互行为。从另一个方面看，如果你认为活动状态和打开状态是有重叠的部分，那么控件的值可能会改变，但是被选中的选项肯定不会相应的进行突出显示，同样是因为我们没有定义在控件打开状态下的任何键盘交互事件（我们仅仅定义了控件打开会发生什么，而没有定义在其打开后会发生什么）
+在我们的示例中，缺失的规范是显而易见的，所以我们将着手处理它们，但是对于一些没有人想到去定义正确行为的控件而言，这的确是一个问题。所以在元素（例如 {{htmlelement('select')}}）标准化阶段，规范作者花费了大量的时间来定义每个输入设备每个用例的所有涉及的交互。创建新的控件并不容易，特别是你正在创建以前从未做过的东西，没有人知道其预期的行为和相关的交互是什么。至少 select 已经完成了这些设计，所以我们知道它应该如何表现！
 
-在我们的例子中，缺失的规范是显而易见的，所以我们将着手处理他们，但是对于一些没有人想到去定义正确行为的小部件而言，这的确是一个问题。所以在设计阶段花费时间是值得的，因为如果你定义的行为不够好，或者忘记定义了一个行为，那么在用户开始实际使用时，将会很难去重新定义它们。如果你在定义时有疑问，请征询他人的意见，如果你有预算，请不要犹豫的去进行[用户可用性测试](https://zh.wikipedia.org/wiki/可用性测试)，这个过程被称为用户体验设计（UX Design），如果你想要深入的学习相关的内容，请查阅下面这些有用资源：
+设计新的交互方式只是行业中重要参与者的一种选择，他们有足够的影响力来推动它们创建的交互方式成为标准。例如，Apple 于 2001 年在 iPod 中推出了滚轮。他们拥有足够的市场份额，而成功推出了一种全新的设备交互方式，这是大多数设备公司无法做到的。
 
-- [UXMatters.com](http://www.uxmatters.com/)
+最好不要发明新的用户交互方式。对于你添加的任何交互方式，在设计阶段花费时间至关重要；如果你对一种行为的定义不够合适，，或者忘记定义了某种行为，那么在用户习惯之后，将很难去重新定义它们。如果你在定义时有疑问，请征询他人的意见，如果你有预算，请不要犹豫去进行[用户可用性测试](https://zh.wikipedia.org/wiki/可用性测试)，这个过程被称为用户体验设计（UX Design），如果你想要深入的学习相关的内容，请查阅下面这些有用资源：
```
Comment (jasonren0403) [inline files/zh-cn/learn/forms/how_to_build_custom_form_controls/index.md:67]:
```suggestion
最好不要发明新的用户交互方式。对于你添加的任何交互方式，在设计阶段花费时间至关重要；如果你对一种行为的定义不够合适，或者忘记定义了某种行为，那么在用户习惯之后，将很难去重新定义它们。如果你在定义时有疑问，请征询他人的意见，如果你有预算，请不要犹豫去进行[用户可用性测试](https://zh.wikipedia.org/wiki/可用性测试)，这个过程被称为用户体验设计（UX Design），如果你想要深入的学习相关的内容，请查阅下面这些有用资源：
```

File: files/zh-cn/learn/forms/how_to_build_custom_form_controls/index.md [hunk 2]
```diff
@@ -5,80 +5,87 @@ slug: Learn/Forms/How_to_build_custom_form_controls
...
+
+- [UXMatters.com](https://www.uxmatters.com/)
 - [UXDesign.com](http://uxdesign.com/)
```
Comment (jasonren0403) [inline files/zh-cn/learn/forms/how_to_build_custom_form_controls/index.md:70]: 用https链接更好些吧 🤔 

```suggestion
- [UXDesign.com](https://uxdesign.com/)
```

File: files/zh-cn/learn/forms/how_to_build_custom_form_controls/index.md [hunk 3]
```diff
@@ -91,34 +98,36 @@ slug: Learn/Forms/How_to_build_custom_form_controls
 </div>
 ```
 
-注意类名的使用：不管实际使用了哪种底层 HTML 元素，它们都标识每个相关的部分。这很重要，因为这样做能确保我们的 CSS 和 JavaScript 不会和 HTML 结构强绑定，这样我们就可以在不破坏使用小部件的代码的情况下进行实现更改。比如，如果你希望增加一个等价的{{HTMLElement("optgroup")}}元素。
+注意类名的使用：不管实际使用了哪种底层 HTML 元素，它们都标识每个相关的部分。这很重要，因为这样做能确保我们的 CSS 和 JavaScript 不会和 HTML 结构强绑定，这样我们就可以在不破坏使用控件的代码的情况下进行实现更改。比如，如果你希望增加一个等价的 {{HTMLElement("optgroup")}} 元素。
```
Comment (jasonren0403) [inline files/zh-cn/learn/forms/how_to_build_custom_form_controls/index.md:102]: 这里少了一段

File: files/zh-cn/learn/forms/how_to_build_custom_form_controls/index.md [hunk 4]
```diff
@@ -142,78 +151,71 @@ slug: Learn/Forms/How_to_build_custom_form_controls
...
-  /* 这是一个以可访问形式隐藏列表的简单方法，
+  /* 这是一个以无障碍的形式隐藏列表的简单方法，
      对无障碍我们将在最后进一步拓展 */
   max-height: 0;
   visibility: hidden;
 }
 ```
 
-#### 美化
+> **备注：** 我们也可以使用 `transform: scale(1, 0)` 来不指定选项列表的高度为零但宽度不变。
```
Comment (jasonren0403) [inline files/zh-cn/learn/forms/how_to_build_custom_form_controls/index.md:161]:
```suggestion
> **备注：** 我们也可以使用 `transform: scale(1, 0)` 来指定选项列表的高度为零，但宽度不变。
```

File: files/zh-cn/learn/forms/how_to_build_custom_form_controls/index.md [hunk 5]
```diff
@@ -142,78 +151,71 @@ slug: Learn/Forms/How_to_build_custom_form_controls
...
-  border-radius: 0.4em; /* 4px */
-  box-shadow: 0 0.1em 0.2em rgba(0, 0, 0, 0.45); /* 0 1px 2px */
+  border: 0.2em solid #000;
+  border-radius: 0.4em;
+  box-shadow: 0 0.1em 0.2em rgba(0, 0, 0, 0.45);
 
-  /* 第一段声明是为了不支持线性梯度填充的浏览器准备的。
-     第二段声明是因为基于 WebKit 的浏览器没有预先定义它。
-     如果你想为过时的浏览器提供支持，请参阅 http://www.colorzilla.com/gradient-editor/ */
+  /* 第一段声明是为不支持线性梯度的浏览器准备的。 */
```
Comment (jasonren0403) [inline files/zh-cn/learn/forms/how_to_build_custom_form_controls/index.md:184]:
```suggestion
  /* 第一段声明是为不支持线性渐变的浏览器准备的。 */
```

File: files/zh-cn/learn/forms/how_to_build_custom_form_controls/index.md [hunk 6]
```diff
@@ -142,78 +151,71 @@ slug: Learn/Forms/How_to_build_custom_form_controls
...
   background: #f0f0f0;
   background: -webkit-linear-gradient(90deg, #e3e3e3, #fcfcfc 50%, #f0f0f0);
```
Comment (jasonren0403) [inline files/zh-cn/learn/forms/how_to_build_custom_form_controls/index.md:186]:
```suggestion
```

File: files/zh-cn/learn/forms/how_to_build_custom_form_controls/index.md [hunk 7]
```diff
@@ -142,78 +151,71 @@ slug: Learn/Forms/How_to_build_custom_form_controls
...
   background: linear-gradient(0deg, #e3e3e3, #fcfcfc 50%, #f0f0f0);
 }
 
 .select .value {
-  /* 因为值的宽度可能超过组件的宽度，我们需要确保他不会改变组件的宽度 */
+  /* 因为值的宽度可能超过控件的宽度，我们需要确保它不会改变空间的宽度。如果内容溢出了，我们显示省略号 */
```
Comment (jasonren0403) [inline files/zh-cn/learn/forms/how_to_build_custom_form_controls/index.md:191]:
```suggestion
  /* 因为值的宽度可能超过控件的宽度，我们需要确保它不会改变控件的宽度。如果内容溢出了，我们显示省略号 */
```

File: files/zh-cn/learn/forms/how_to_build_custom_form_controls/index.md [hunk 8]
```diff
@@ -142,78 +151,71 @@ slug: Learn/Forms/How_to_build_custom_form_controls
...
-
-  /* 如果内容溢出了，最好有一个恰当的缩写。*/
   white-space: nowrap;
   text-overflow: ellipsis;
+  vertical-align: top;
 }
 ```
 
-我们不需要一个额外的元素来设计向下的箭头，而使用{{cssxref(":after")}} 伪类来替代。然而，这也可以通过使用一张加在`select` class 上的简单的背景图像来实现。
+我们不需要一个额外的元素来设计向下箭头，而使用 {{cssxref(":after")}} 伪类替代。这也可以通过使用一张加在`select` 类上的简单的背景图像来实现。
```
Comment (jasonren0403) [inline files/zh-cn/learn/forms/how_to_build_custom_form_controls/index.md:201]:
```suggestion
我们不需要一个额外的元素来设计向下箭头，而使用 {{cssxref(":after")}} 伪类替代。这也可以通过使用一张加在 `select` 类上的简单的背景图像来实现。
```

File: files/zh-cn/learn/forms/how_to_build_custom_form_controls/index.md [hunk 9]
```diff
@@ -225,35 +227,34 @@ slug: Learn/Forms/How_to_build_custom_form_controls
...
   /* 这会重置 ul 元素的默认样式 */
   list-style: none;
   margin: 0;
   padding: 0;
 
-  -moz-box-sizing: border-box;
   box-sizing: border-box;
 
-  /* 这会确保即使数值比组件小，选项列表仍能变得跟组件自身一样大*/
+  /* 这会确保即使数值比控件小，选项列表仍能变得跟控件自身一样大 */
```
Comment (jasonren0403) [inline files/zh-cn/learn/forms/how_to_build_custom_form_controls/index.md:239]:
```suggestion
  /* 这会确保即使数值比控件小，选项列表仍能变得跟控件自身一样宽 */
```

File: files/zh-cn/learn/forms/how_to_build_custom_form_controls/index.md [hunk 10]
```diff
@@ -266,363 +267,1353 @@ slug: Learn/Forms/How_to_build_custom_form_controls
...
+
+## 通过 JavaScript 让你的控件动起来
+
+现在我们的设计和结构已经完成了。我们可以写些 JavaScript 代码来让这个控件真正生效。
+
+> **警告：** 下面的代码仅仅是教学性质的，不是生产环境的代码，并且不应该照搬使用。这种方案不具有前瞻性，而且可能在旧式浏览器上会不工作。这里面还有冗余的部分，在生产环境下，代码需要优化。
+
+### 它为什么不生效？
+
+在我们开始之前，要记住一件和 JavaScript 有关的非常重要的事情：**在浏览器中，这是一种不可靠的技术**。当你构建一个自定义控件时，你会不得不得依赖于 JavaScript，因为这是将所有的东西联系在一起的线索。但是，很多情况下，JavaScript 不能在浏览器中运行。
```
Comment (jasonren0403) [inline files/zh-cn/learn/forms/how_to_build_custom_form_controls/index.md:646]:
```suggestion
在我们开始之前，要记住一件和 JavaScript 有关的非常重要的事情：**在浏览器中，这是一种不可靠的技术**。当你构建一个自定义控件时，你会不得不依赖于 JavaScript，因为这是将所有的东西联系在一起的线索。但是，很多情况下，JavaScript 不能在浏览器中运行。
```

File: files/zh-cn/learn/forms/how_to_build_custom_form_controls/index.md [hunk 11]
```diff
@@ -701,21 +1688,382 @@ function updateValue(select, index) {
...
+
+我们可以从一个完全语义化、无障碍且无序的{{htmlelement('input/radio','单选')}}按钮列表开始，使用一对语义化的 {{htmlelement('fieldset')}} 和 {{htmlelement('legend')}} 对来标记整个组。
+
+```html
+<fieldset>
+  <legend>Pick a fruit</legend>
+  <ul class="styledSelect">
+    <li><input type="radio" name="fruit" value="Cherry" id="fruitCherry" checked><label for="fruitCherry">Cherry</label></li>
+    <li><input type="radio" name="fruit" value="Lemon" id="fruitLemon"><label for="fruitLemon">Lemon</label></li>
+    <li><input type="radio" name="fruit" value="Banana" id="fruitBanana"><label for="fruitBanana"">Banana</label></li>
```
Comment (jasonren0403) [inline files/zh-cn/learn/forms/how_to_build_custom_form_controls/index.md:1968]:
```suggestion
    <li><input type="radio" name="fruit" value="Banana" id="fruitBanana"><label for="fruitBanana">Banana</label></li>
```

**Review by jasonren0403 — APPROVED** [review]
👍 Thank you

**Pull request merged**